pageserver/src/tenant.rs

```diff
@@ -1398,52 +1462,191 @@ impl Tenant {
         }
     }
 
+    fn check_to_be_archived_has_no_unarchived_children(
+        timeline_id: TimelineId,
+        timelines: &std::sync::MutexGuard<'_, HashMap<TimelineId, Arc<Timeline>>>,
+    ) -> Result<(), TimelineArchivalError> {
+        let children: Vec<TimelineId> = timelines
+            .iter()
+            .filter_map(|(id, entry)| {
+                if entry.get_ancestor_timeline_id() != Some(timeline_id) {
+                    return None;
+                }
+                if entry.is_archived() == Some(true) {
+                    return None;
+                }
+                Some(*id)
+            })
+            .collect();
+
+        if !children.is_empty() {
+            return Err(TimelineArchivalError::HasUnarchivedChildren(children));
+        }
+        Ok(())
+    }
+
+    fn check_ancestor_of_to_be_unarchived_is_not_archived(
+        timeline_id: TimelineId,
+        ancestor_timeline_id: TimelineId,
+        timelines: &std::sync::MutexGuard<'_, HashMap<TimelineId, Arc<Timeline>>>,
+        offloaded_timelines: &std::sync::MutexGuard<
+            '_,
+            HashMap<TimelineId, Arc<OffloadedTimeline>>,
+        >,
+    ) -> Result<(), TimelineArchivalError> {
+        let has_archived_parent = if let Some(ancestor_timeline) = timelines.get(&timeline_id) {
+            ancestor_timeline.is_archived() == Some(true)
+        } else if offloaded_timelines.contains_key(&ancestor_timeline_id) {
+            true
+        } else {
+            error!("ancestor timeline {ancestor_timeline_id} not found");
+            debug_assert!(false);
+            return Err(TimelineArchivalError::NotFound);
+        };
+        if has_archived_parent {
+            return Err(TimelineArchivalError::HasArchivedParent(
+                ancestor_timeline_id,
+            ));
+        }
+        Ok(())
+    }
+
+    fn check_to_be_unarchived_timeline_has_no_archived_parent(
```

Comment:
I expected this function to loop through all ancestors.

---

Also, the check is different from `check_to_be_archived_has_no_unarchived_children` which also looks inside `offloaded_timelines`. That's a bit weird.

Reply:
I think it makes no sense to look in `offloaded_timelines` because we already have the `Timeline` object.

I'll add code to prevent offloading if there is *any* children that are not offloaded. Then, conversely, we know that if a timeline isn't offloaded, its ancestor isn't either (archived, but not offloaded).

Reply:
ahh it's already there, it's just named wrongly. renamed.

Reply:
Ah I see. One function is doing the lookup and this one already has the timeline object. Maybe it would clearer if this was a method on `Timeline` (takes `self: &Arc<Self>`).

Reply:
yeah, good idea. I'll do a followup.